### DISCUSSION TITLE
Test에서 Bean이 생성되지 않는 이유가 무엇일까요?

### DISCUSSION THREAD

**Discussion by jj1kim** [wafflestudio/seminar-2023#193] [Q&A] [answered]
안녕하세요. 어느 정도 과제에 대한 구현을 했다고 생각해서 테스트를 돌려보니, 1번 2번 문제에 대한 테스트가 모두 실패하는 것을 확인 했습니다. 이유를 보니, test로 받는 값들에 대해서 bean이 형성이 되지 않았다고 하더군요ㅜㅜ
일단 아래는 구현한 PlaylistViewServiceImpl의 코드입니다.
```
class PlaylistViewServiceImpl(val playlistRepository: PlaylistRepository) : PlaylistViewService, SortPlaylist {
    private val userLastViewedMap = mutableMapOf<Long, LocalDateTime>()
    /**
     * 스펙:
     *  1. 같은 유저-같은 플레이리스트의 조회 수는 1분에 1개까지만 허용한다.
     *  2. PlaylistView row 생성, PlaylistEntity의 viewCnt 업데이트가 atomic하게 동작해야 한다. (둘 다 모두 성공하거나, 둘 다 모두 실패해야 함)
     *  3. 플레이리스트 조회 요청이 동시에 다수 들어와도, 요청이 들어온 만큼 조회 수가 증가해야한다. (동시성 이슈가 없어야 함)
     *  4. 성공하면 true, 실패하면 false 반환
     *
     * 조건:
     *  1. Synchronized 사용 금지.
     *  2. create 함수 처리가, 플레이리스트 조회 API 응답시간에 영향을 미치면 안된다.
     *  3. create 함수가 실패해도, 플레이리스트 조회 API 응답은 성공해야 한다.
     *  4. Future가 리턴 타입인 이유를 고민해보며 구현하기.
     */
    override fun create(playlistId: Long, userId: Long, at: LocalDateTime): Future<Boolean> {
        val playlist = playlistRepository.findById(playlistId)
                .orElseThrow { NoSuchElementException("Playlist not found") }

        val lastViewedTime = userLastViewedMap[userId]

        if (lastViewedTime == null || Duration.between(lastViewedTime, at).toMinutes() >= 1) {
            userLastViewedMap[userId] = at
            playlist.viewCnt++
            playlistRepository.save(playlist)
            return CompletableFuture.completedFuture(true)
        }

        return CompletableFuture.completedFuture(false)
    }

    override fun invoke(playlists: List<PlaylistBrief>, type: Type, at: LocalDateTime): List<PlaylistBrief> {
        return when (type) {
            Type.DEFAULT -> playlists
            else -> TODO("Not yet implemented")
        }
    }
}
```
아직 else 부분은 구현을 안했지만, 이렇게 해도 테스트가 일단 돌아갈 것이라고 생각하고 PlaylistViewServiceTest를 돌렸습니다. 그러나 저 문제로 걸리기는 커녕 테스트 코드의 아래 부분에서 문제가 발생하더군요.
```
@SpringBootTest
class PlaylistViewServiceTest @Autowired constructor(
    private val playlistViewService: PlaylistViewService,
    private val sortPlaylist: SortPlaylist,
    private val playlistRepository: PlaylistRepository,
    private val playlistViewRepository: PlaylistViewRepository,
) {
```
여기서 playlistViewService, sortPlaylist,playlistRepository,playlistViewRepository 각각의 Bean이 생성되지 않아서 테스트 자체가 제대로 실행이 되지 않고 있음을 확인했습니다. 이 경우에는 무엇이 문제일까요? 제가 작성한 로직 때문에 발생하는 문제는 아닌거 같은데...
사실 굉장히 바보같은 질문일 수도 있는데, 이때까지 과제할 때는 이런 일이 없었던지라 굉장히 당황스럽네요... 문제를 해결하려고 이곳저곳 뒤져보았지만 마땅한 답을 찾기 어려워서 질문드립니다...

**Comment by PFCJeong** [comment]
안녕하세요~

1. `PlaylistViewServiceImpl`에 `@Service` 어노테이션이 달린거 맞는지 확인 부탁드립니다.
2. 어노테이션이 달려있다면,, 첨부주신 코드만으로는 확인이 불가능해서 전체 코드를 봐야할 것 같습니다. 지금 작업 중이신 프라이빗 레포에 collaborator로 저를 추가해주시면 확인해보겠습니다.

**Reply by jj1kim** [reply]
프라이빗 레포에 collaborator로 추가했습니다! 확인 부탁드립니다ㅜㅜ

**Reply by PFCJeong** [reply, marked as answer]
`
Caused by: org.hibernate.AnnotationException: Collection 'com.wafflestudio.seminar.spring2023.user.repository.UserEntity.playlistLikes' is 'mappedBy' a property named 'user' which does not exist in the target entity 'com.wafflestudio.seminar.spring2023.playlist.repository.PlaylistLikeEntity'
`

요 에러 메시지가 나오고 있네요. 이거 해결하면 될 것 같습니다.

**Reply by jj1kim** [reply]
문제 찾았습니다ㅜㅜ 매번 세미나 내용을 로컬에서 작업할 때, 세미나 주별로 업데이트 된 내용을 풀 땡겨오다보니 원래 작업했던 코드랑 새로 나온 코드랑 뒤죽박죽 섞여서 제대로 작업이 안되더군요.. 해결했습니다!! 감사합니다!!
